frontend/src/router/pages/SubmitDates.jsx

```diff
@@ -160,32 +166,43 @@ function SubmitDates() {
           },
         ],
       },
-    });
+    }));
   }
 
-  function updateDateRange(dateableId, dateType, index, key, value) {
-    setDates({
-      ...dates,
-      [dateableId]: {
-        ...dates[dateableId],
-        [dateType]: dates[dateableId][dateType].map((dateRange, i) =>
-          i === index
-            ? { ...dateRange, [key]: formatDatetoISO(value), changed: true }
-            : dateRange,
-        ),
-      },
+  function updateDateRange(
+    dateableId,
+    dateType,
+    index,
+    key,
+    value,
+    callback = null,
+  ) {
+    const newValue = value ? value.toISOString() : null;
+
+    setDates((prevDates) => {
+      const updatedDates = cloneDeep(prevDates);
+
+      // Update the date value and mark the date range as changed
+      lodashSet(updatedDates, [dateableId, dateType, index, key], newValue);
+      lodashSet(updatedDates, [dateableId, dateType, index, "changed"], true);
+
+      if (callback) {
+        callback(updatedDates);
+      }
```

Comment:
I think it'd be better to use `useEffect` with the dates dependency. That's pretty much a listener that will be triggered when anything in the dates object changes.

Reply:
I tried a few different approaches but I kept running into trouble with stale state. In this case I wanted to set it up so you could specify the callback (or have no callback at all), but it still feels clunky... I will remember it for refactoring in the future, and hopefully I'll learn a better pattern in the meantime, haha.